casper-finality-proofs/src/justification_bits.rs

```diff
@@ -0,0 +1,81 @@
+use plonky2x::prelude::RichField;
+use plonky2x::{
+    frontend::vars::SSZVariable,
+    prelude::{
+        ArrayVariable, BoolVariable, ByteVariable, Bytes32Variable, CircuitBuilder,
+        CircuitVariable, PlonkParameters, Variable,
+    },
+};
+
+#[derive(Debug, Clone, CircuitVariable)]
+#[value_name(JustificationBitsValue)]
+pub struct JustificationBitsVariable {
+    pub bits: ArrayVariable<BoolVariable, 4>,
+}
+
+impl JustificationBitsVariable {
+    pub fn test_range<L: PlonkParameters<D>, const D: usize>(
+        &self,
+        builder: &mut CircuitBuilder<L, D>,
+        lower_bound: usize,
+        upper_bound_non_inclusive: usize,
+    ) -> BoolVariable {
+        let mut result = builder._true();
+        for i in lower_bound..upper_bound_non_inclusive {
+            result = builder.and(result, self.bits[i]);
+        }
+        result
+    }
+
+    pub fn shift_right<L: PlonkParameters<D>, const D: usize>(
+        &self,
+        builder: &mut CircuitBuilder<L, D>,
+    ) -> JustificationBitsVariable {
+        let mut new_bits = self.bits.as_vec();
+        for i in (0..3).rev() {
+            new_bits[i + 1] = new_bits[i];
+        }
+        new_bits[0] = builder._false();
```

Comment:
Please make it generic in a sense that it can be reused by other circuits (pass a slice and shift count)